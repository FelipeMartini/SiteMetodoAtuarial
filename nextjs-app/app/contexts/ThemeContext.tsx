/**
 * Contexto de Tema Unificado
 * Sistema moderno de gerenciamento de temas para a aplicação
 */
'use client';

import React, { createContext, useContext, useState, useEffect, ReactNode } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { Theme, ThemeName, getThemeByName, getNextTheme, lightTheme } from '../../styles/themes';

interface ThemeContextProps {
  currentTheme: Theme;
  themeName: ThemeName;
  toggleTheme: () => void;
  setTheme: (name: ThemeName) => void;
  isDarkMode: boolean;
}

const ThemeContext = createContext<ThemeContextProps | undefined>(undefined);

interface ThemeProviderProps {
  children: ReactNode;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const [themeName, setThemeName] = useState<ThemeName>('light');
  const [currentTheme, setCurrentTheme] = useState<Theme>(lightTheme);

  // Carrega tema do localStorage na inicialização
  useEffect(() => {
    const savedTheme = localStorage.getItem('theme') as ThemeName;
    if (savedTheme && (savedTheme === 'light' || savedTheme === 'dark')) {
      setThemeName(savedTheme);
      setCurrentTheme(getThemeByName(savedTheme));
    }
  }, []);

  // Salva tema no localStorage quando muda
  useEffect(() => {
    localStorage.setItem('theme', themeName);
    setCurrentTheme(getThemeByName(themeName));
  }, [themeName]);

  const toggleTheme = () => {
    const nextTheme = getNextTheme(themeName);
    setThemeName(nextTheme.name as ThemeName);
  };

  const setTheme = (name: ThemeName) => {
    setThemeName(name);
  };

  const isDarkMode = themeName === 'dark';

  const value: ThemeContextProps = {
    currentTheme,
    themeName,
    toggleTheme,
    setTheme,
    isDarkMode,
  };

  return (
    <ThemeContext.Provider value={value}>
      <StyledThemeProvider theme={currentTheme}>
        {children}
      </StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

export const useTheme = (): ThemeContextProps => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme deve ser usado dentro de um ThemeProvider');
  }
  return context;
};

// Alias para compatibilidade com código legado
export const useTema = useTheme;
export const useUtilsTema = useTheme;
