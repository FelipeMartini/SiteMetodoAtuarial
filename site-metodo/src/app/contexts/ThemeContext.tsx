// Arquivo movido para @core/theme/ContextoTema.tsx
// Remover este arquivo se não houver dependências legadas.
