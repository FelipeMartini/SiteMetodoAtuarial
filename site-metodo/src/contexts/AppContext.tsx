// Exemplo de AppContext
import { createContext } from "react";
export const AppContext = createContext({});
