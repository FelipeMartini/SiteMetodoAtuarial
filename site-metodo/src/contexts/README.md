# Contextos Globais

- Contextos de autenticação, tema, permissões, etc.
