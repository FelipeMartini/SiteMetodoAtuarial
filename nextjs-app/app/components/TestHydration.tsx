// TestHydration.tsx - Componente para testar se hydration est√° funcionando
'use client';

import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { useTema } from '../contexts/ThemeContext';

// Styled component para testar
const TesteContainer = styled.div`
  padding: ${props => props.theme.espacamentos.lg};
  background: ${props => props.theme.colors.backgroundSecondary};
  border: 2px solid ${props => props.theme.colors.primary};
  border-radius: ${props => props.theme.bordas.raios.lg};
  margin: ${props => props.theme.espacamentos.md} 0;
  transition: all ${props => props.theme.animacoes.transicoes.normal};

  &:hover {
    background: ${props => props.theme.colors.surface};
    transform: translateY(-2px);
    box-shadow: ${props => props.theme.sombras.lg};
  }
`;

const TituleTeste = styled.h3`
  color: ${props => props.theme.colors.primary};
  font-size: ${props => props.theme.tipografia.tamanhos.xl};
  font-weight: ${props => props.theme.tipografia.pesos.negrito};
  margin-bottom: ${props => props.theme.espacamentos.sm};
`;

const TextoTeste = styled.p`
  color: ${props => props.theme.colors.textSecondary};
  font-size: ${props => props.theme.tipografia.tamanhos.base};
  line-height: ${props => props.theme.tipografia.alturas.relaxed};
  margin-bottom: ${props => props.theme.espacamentos.md};
`;

const BotaoTeste = styled.button`
  background: ${props => props.theme.colors.primary};
  color: ${props => props.theme.colors.textInverted};
  border: none;
  border-radius: ${props => props.theme.bordas.raios.md};
  padding: ${props => props.theme.espacamentos.sm} ${props => props.theme.espacamentos.md};
  font-size: ${props => props.theme.tipografia.tamanhos.base};
  font-weight: ${props => props.theme.tipografia.pesos.medio};
  cursor: pointer;
  transition: all ${props => props.theme.animacoes.transicoes.rapida};

  &:hover {
    background: ${props => props.theme.colors.primaryHover};
    transform: translateY(-1px);
  }

  &:active {
    transform: translateY(0);
  }
`;

const StatusIndicator = styled.div<{ $status: 'loading' | 'success' | 'errorr' }>`
  display: inline-block;
  padding: ${props => props.theme.espacamentos.xs} ${props => props.theme.espacamentos.sm};
  border-radius: ${props => props.theme.bordas.raios.full};
  font-size: ${props => props.theme.tipografia.tamanhos.sm};
  font-weight: ${props => props.theme.tipografia.pesos.medio};
  ${props => {
    switch (props.$status) {
      case 'loading':
        return `background: ${props.theme.colors.warningBackground}; color: ${props.theme.colors.warning}; border: 1px solid ${props.theme.colors.warning};`;
      case 'success':
        return `background: ${props.theme.colors.successBackground}; color: ${props.theme.colors.success}; border: 1px solid ${props.theme.colors.success};`;
      case 'errorr':
        return `background: ${props.theme.colors.errorrBackground}; color: ${props.theme.colors.error}; border: 1px solid ${props.theme.colors.error};`;
      default:
        return '';
    }
  }}
`;

export default function TestHydration() {
  const { temaAtual, alternarTema, isHydrated } = useTema();
  const [renderTime, setRenderTime] = useState<string>('');
  const [clickCount, setClickCount] = useState(0);

  useEffect(() => {
    setRenderTime(new Date().toLocaleTimeString());
  }, []);

  const handleTemaClick = () => {
    alternarTema();
    setClickCount(prev => prev + 1);
  };

  const getHydrationStatus = () => {
    if (!isHydrated) return 'loading';
    return renderTime ? 'success' : 'errorr';
  };

  const getStatusText = () => {
    const status = getHydrationStatus();
    switch (status) {
      case 'loading':
        return '‚è≥ Carregando...';
      case 'success':
        return '‚úÖ Hydrated';
      case 'errorr':
        return '‚ùå Erro';
      default:
        return '';
    }
  };

  return (
    <TesteContainer>
      <TituleTeste>üß™ Teste de Hydration - Styled Components</TituleTeste>
      <TextoTeste>
        Este componente testa se o styled-components est√° funcionando corretamente
        no Next.js 15 com Server Side Rendering e sem errors de hydration.
      </TextoTeste>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Status do Hydration:</strong>{' '}
        <StatusIndicator $status={getHydrationStatus()}>
          {getStatusText()}
        </StatusIndicator>
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Tema Atual:</strong> {temaAtual.displayName} {temaAtual.icone}
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Renderizado em:</strong> {renderTime || 'Aguardando...'}
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Cliques no bot√£o:</strong> {clickCount}
      </div>
      <BotaoTeste onClick={handleTemaClick}>
        üé® Alternar Tema ({clickCount} cliques)
      </BotaoTeste>
      <div style={{ marginTop: '1rem', fontSize: '0.875rem', opacity: 0.7 }}>
        üí° <strong>Como testar:</strong>
        <br />
        1. Abra o DevTools (F12)
        <br />
        2. Recarregue a p√°gina (F5)
        <br />
        3. Verifique se n√£o h√° errors de hydration no console
        <br />
        4. Observe se os estilos carregam instantaneamente (sem flicker)
        <br />
        5. Teste a altern√¢ncia de temas clicando no bot√£o
      </div>
    </TesteContainer>
  );
}

// Coment√°rio: Este componente serve para validar visualmente e funcionalmente o SSR do styled-components, altern√¢ncia de temas e status de hydration. Deve ser usado em testes automatizados e manuais.
