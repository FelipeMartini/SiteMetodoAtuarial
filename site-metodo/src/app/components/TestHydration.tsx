// TestHydration.tsx - Componente para testar se hydration est√° funcionando
'use client';

import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import { useTema } from '../contexts/ThemeContext';

// Styled component para testar
const TesteContainer = styled.div`
  padding: ${({ theme }) => theme.spacing.lg};
  background: ${({ theme }) => theme.colors.backgroundSecondary};
  border: 2px solid ${({ theme }) => theme.colors.primary};
  border-radius: ${({ theme }) => theme.borderRadius.lg};
  margin: ${({ theme }) => theme.spacing.md} 0;
  transition: all ${({ theme }) => theme.transitions.normal};
  &:hover {
    background: ${({ theme }) => theme.colors.surface};
    transform: translateY(-2px);
    box-shadow: ${({ theme }) => theme.shadows.lg};
  }
`;

const TituleTeste = styled.h3`
  color: ${({ theme }) => theme.colors.primary};
  font-size: ${({ theme }) => theme.typography.fontSize.xl};
  font-weight: ${({ theme }) => theme.typography.fontWeight.bold};
  margin-bottom: ${({ theme }) => theme.spacing.sm};
`;

const TextoTeste = styled.p`
  color: ${({ theme }) => theme.colors.textSecondary};
  font-size: ${({ theme }) => theme.typography.fontSize.base};
  line-height: ${({ theme }) => theme.typography.lineHeight.relaxed};
  margin-bottom: ${({ theme }) => theme.spacing.md};
`;

// ...styled do BotaoTeste j√° est√° definido corretamente abaixo...
const BotaoTeste = styled.button`
  background: ${({ theme }) => theme.colors.primary};
  color: ${({ theme }) => theme.colors.textInverted};
  border: none;
  border-radius: ${({ theme }) => theme.borderRadius.md};
  padding: ${({ theme }) => theme.spacing.sm} ${({ theme }) => theme.spacing.md};
  font-size: ${({ theme }) => theme.typography.fontSize.base};
  font-weight: ${({ theme }) => theme.typography.fontWeight.medium};
  cursor: pointer;
  transition: all ${({ theme }) => theme.transitions.fast};
  &:hover {
    background: ${({ theme }) => theme.colors.primaryHover};
    transform: translateY(-1px);
  }
  &:active {
    transform: translateY(0);
  }
`;

// ...styled do StatusIndicator j√° est√° definido corretamente abaixo...
const StatusIndicator = styled.div<{ $status: 'success' | 'errorr' }>`
  display: inline-block;
  padding: ${({ theme }) => theme.spacing.xs} ${({ theme }) => theme.spacing.sm};
  border-radius: ${({ theme }) => theme.borderRadius.full};
  font-size: ${({ theme }) => theme.typography.fontSize.sm};
  font-weight: ${({ theme }) => theme.typography.fontWeight.medium};
  ${({ $status, theme }) => {
    switch ($status) {
      case 'success':
        return `background: ${theme.colors.successBackground}; color: ${theme.colors.success}; border: 1px solid ${theme.colors.success};`;
      case 'errorr':
        return `background: ${theme.colors.errorBackground}; color: ${theme.colors.error}; border: 1px solid ${theme.colors.error};`;
      default:
        return '';
    }
  }}
`;

interface TestHydrationProps {
  dataRenderizacao?: Date;
}

export default function TestHydration({ dataRenderizacao }: TestHydrationProps) {
  // Hook do tema e estados locais
  const { currentTheme, toggleTheme } = useTema();
  const [renderTime, setRenderTime] = useState<string>('');
  const [clickCount, setClickCount] = useState(0);

  // Marca o hor√°rio de renderiza√ß√£o para testar hydration
  useEffect(() => {
    const data = dataRenderizacao || new Date();
    setRenderTime(data.toLocaleTimeString());
  }, [dataRenderizacao]);

  // Alterna o tema e conta cliques
  const handleTemaClick = () => {
    toggleTheme();
    setClickCount(prev => prev + 1);
  };

  // Status de hydration
  const getHydrationStatus = () => {
    return renderTime ? 'success' : 'errorr';
  };

  // Texto do status
  const getStatusText = () => {
    const status = getHydrationStatus();
    switch (status) {
      case 'success':
        return '‚úÖ Hydrated';
      case 'errorr':
        return '‚ùå Erro';
      default:
        return '';
    }
  };

  // Renderiza√ß√£o principal do componente
  return (
    <TesteContainer>
      <TituleTeste>Teste de Hydration</TituleTeste>
      <TextoTeste>
        Este componente valida visualmente e funcionalmente o SSR do styled-components, altern√¢ncia de temas e status de hydration.
      </TextoTeste>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Tema Atual:</strong> {currentTheme?.name ? currentTheme.name.charAt(0).toUpperCase() + currentTheme.name.slice(1) : 'Desconhecido'}
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Renderizado em:</strong> {renderTime || 'Aguardando...'}
      </div>
      <div style={{ marginBottom: '1rem' }}>
        <strong>Cliques no bot√£o:</strong> {clickCount}
      </div>
      <BotaoTeste onClick={handleTemaClick}>
        üé® Alternar Tema ({clickCount} cliques)
      </BotaoTeste>
      <div style={{ margin: '1rem 0' }}>
        <StatusIndicator $status={getHydrationStatus()}>
          {getStatusText()}
        </StatusIndicator>
      </div>
      <div style={{ marginTop: '1rem', fontSize: '0.875rem', opacity: 0.7 }}>
        üí° <strong>Como testar:</strong>
        <br />
        1. Abra o DevTools (F12)
        <br />
        2. Recarregue a p√°gina (F5)
        <br />
        3. Verifique se n√£o h√° erros de hydration no console
        <br />
        4. Observe se os estilos carregam instantaneamente (sem flicker)
        <br />
        5. Teste a altern√¢ncia de temas clicando no bot√£o
      </div>
    </TesteContainer>
  );
}

// Coment√°rio: Este componente serve para validar visualmente e funcionalmente o SSR do styled-components, altern√¢ncia de temas e status de hydration. Deve ser usado em testes automatizados e manuais.
