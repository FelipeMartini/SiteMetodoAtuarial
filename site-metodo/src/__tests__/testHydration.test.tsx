// Teste automatizado para o componente TestHydration
// Garante que o componente renderiza corretamente, exibe status de hydration e alterna tema sem erros

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ThemeProvider } from '../app/contexts/ThemeContext';
import TestHydration from '../app/components/TestHydration';

describe('TestHydration', () => {
  it('deve renderizar o componente, exibir status de hydration e bater snapshot', () => {
    // Mock da data para garantir snapshot estável
    const fixedTimestamp = new Date('2025-08-05T12:00:00').getTime();
    jest.spyOn(Date, 'now').mockReturnValue(fixedTimestamp);
    const dataFixa = new Date('2025-08-05T12:00:00');
    const { getByText, getAllByText, asFragment } = render(
      <ThemeProvider>
        <TestHydration dataRenderizacao={dataFixa} />
      </ThemeProvider>
    );
    expect(getByText(/Teste de Hydration/i)).toBeInTheDocument();
    const hydratedElements = getAllByText(/Hydrated|Erro/i);
    expect(hydratedElements.length).toBeGreaterThan(0);
    expect(getByText(/Renderizado em:/i)).toBeInTheDocument();
    expect(getByText(/Cliques no botão:/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
    jest.restoreAllMocks();
  });

  it('deve alternar o tema ao clicar no botão', () => {
    const { getByText } = render(
      <ThemeProvider>
        <TestHydration />
      </ThemeProvider>
    );
    const botao = getByText(/Alternar Tema/i);
    fireEvent.click(botao);
    expect(botao.textContent).toMatch(/1 cliques?/i);
  });
});

// Comentários:
// - Este teste cobre a renderização, status de hydration e alternância de tema do componente TestHydration.
// - Garante que o sistema de temas e styled-components SSR estão funcionando sem erros visuais ou de contexto.
// - Se este teste passar, a adaptação do contexto de tema e SSR está validada para o componente de teste visual.
