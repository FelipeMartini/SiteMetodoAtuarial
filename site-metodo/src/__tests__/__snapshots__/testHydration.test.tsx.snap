// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TestHydration deve renderizar o componente, exibir status de hydration e bater snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bRKDuR jmeLGB"
  >
    <h3
      class="sc-hvigdm kCIdWo"
    >
      Teste de Hydration
    </h3>
    <p
      class="sc-fhHczv gVEFdi"
    >
      Este componente valida visualmente e funcionalmente o SSR do styled-components, altern√¢ncia de temas e status de hydration.
    </p>
    <div
      style="margin-bottom: 1rem;"
    >
      <strong>
        Tema Atual:
      </strong>
       Light
    </div>
    <div
      style="margin-bottom: 1rem;"
    >
      <strong>
        Renderizado em:
      </strong>
       12:00:00
    </div>
    <div
      style="margin-bottom: 1rem;"
    >
      <strong>
        Cliques no bot√£o:
      </strong>
       0
    </div>
    <button
      class="sc-ggWZvA RelyL"
    >
      üé® Alternar Tema (0 cliques)
    </button>
    <div
      style="margin: 1rem 0px;"
    >
      <div
        class="sc-dTvVRJ jrWXks"
      >
        ‚úÖ Hydrated
      </div>
    </div>
    <div
      style="margin-top: 1rem; font-size: 0.875rem; opacity: 0.7;"
    >
      üí° 
      <strong>
        Como testar:
      </strong>
      <br />
      1. Abra o DevTools (F12)
      <br />
      2. Recarregue a p√°gina (F5)
      <br />
      3. Verifique se n√£o h√° erros de hydration no console
      <br />
      4. Observe se os estilos carregam instantaneamente (sem flicker)
      <br />
      5. Teste a altern√¢ncia de temas clicando no bot√£o
    </div>
  </div>
</DocumentFragment>
`;
